src/ring/middleware/session_timeout.clj

```diff
@@ -4,6 +4,40 @@
 (defn- current-time []
   (quot (System/currentTimeMillis) 1000))
 
+(defn- session-expired?
+  [request]
+  (let [session (:session request {})
+        end-time (or (::absolute-timeout session)
+                     (::idle-timeout session))]
+    (and end-time (< end-time (current-time)))))
+
+(defn- wrap-session-timeout
+  [response-fn handler {:keys [timeout timeout-response timeout-handler]
+                        :or {timeout 600}}]
+  (fn
+    ([request]
+     (if (session-expired? request)
+       (assoc (or timeout-response (timeout-handler request)) :session nil)
+       (response-fn request (handler request) timeout)))
+    ([request respond raise]
+     (if (session-expired? request)
+       (if timeout-response
+         (respond (assoc timeout-response :session nil))
+         (timeout-handler request (comp respond #(assoc % :session nil)) raise))
+       (handler request
+                #(respond (response-fn request % timeout))
+                raise)))))
+
+(defn- idle-session-timeout-response
+  [request response timeout]
+  (when response
```

Comment:
Rather than passing in the request, we can just pass in the session, since we already need it to check for expiration (see my previous example code).

Reply:
When we check to see if the session is expired we only look in the request for session data, when we are generating the response we check the response and then the request for the session data. In the original implementation you can see where we fetch the session twice.

```clojure
  (fn [request]
    (let [session  (:session request {})
          end-time (::idle-timeout session)]
      (if (and end-time (< end-time (current-time)))
        (assoc (or timeout-response (timeout-handler request)) :session nil)
        (when-let [response (handler request)]
          (let [session (:session response session)]
            (if (nil? session)
              response
              (let [end-time (+ (current-time) timeout)]
                (assoc response :session (assoc session ::idle-timeout end-time)))))))))
```

I move the `let` for the session closer to response function so that we have the response in-hand, this looks a little clunkier to me than you might have been expecting.

Reply:
Yes, but we don't need the whole request, right? Just the session of the request.